instrumentor/controllers/instrumentedapplication_controller.go

```diff
@@ -0,0 +1,362 @@
+	// Read detection result
+	result := containerStatus.State.Terminated.Message
+	var detectionResult common.DetectionResult
+	err := json.Unmarshal([]byte(result), &detectionResult)
+	if err != nil {
+		logger.Error(err, "error parsing detection result")
+		return err
+	} else {
+		err = r.Get(ctx, namespacedName, &instrumentedApp)
+		if err != nil {
+			logger.Error(err, "error fetching instrumented application object")
+			return err
+		}
+		logger.V(0).Info("detection result", "result", detectionResult)
+		instrumentedApp.Spec.Languages = detectionResult.LanguageByContainer
+		instrumentedApp.Spec.Applications = detectionResult.ApplicationByContainer
+
+		err = r.Update(ctx, &instrumentedApp)
+		if err != nil {
+			logger.Error(err, "error updating InstrumentedApp object with detection result")
+			return err
+		}
+
+		instrumentedApp.Status.InstrumentationDetection.Phase = v1.CompletedInstrumentationDetectionPhase
+		err = r.Status().Update(ctx, &instrumentedApp)
+		if err != nil {
+			logger.Error(err, "error updating InstrumentedApp phase status with detection result")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *InstrumentedApplicationReconciler) startDetection(ctx context.Context, logger logr.Logger, instrumentedApp v1.InstrumentedApplication) (ctrl.Result, error) {
+	instrumentedApp.Status.InstrumentationDetection.Phase = v1.RunningInstrumentationDetectionPhase
+	err := r.Status().Update(ctx, &instrumentedApp)
+	if err != nil {
+		logger.Error(err, "error updating instrument app status")
+		return ctrl.Result{}, err
+	}
+
+	labels, err := r.getOwnerTemplateLabels(ctx, &instrumentedApp)
+	if err != nil {
+		logger.Error(err, "error getting owner labels")
+		return ctrl.Result{}, err
+	}
+
+	err = r.detectLanguage(ctx, &instrumentedApp, labels)
+	if err != nil {
+		logger.Error(err, "error detecting language")
+	}
+	return ctrl.Result{}, err
+}
+
+func (r *InstrumentedApplicationReconciler) shouldStartDetection(app *v1.InstrumentedApplication) bool {
+	return app.Status.InstrumentationDetection.Phase == v1.PendingInstrumentationDetectionPhase
+}
+
+func (r *InstrumentedApplicationReconciler) isLangDetected(app *v1.InstrumentedApplication) bool {
+	return len(app.Spec.Languages) > 0
+}
+
+func (r *InstrumentedApplicationReconciler) isAppDetected(app *v1.InstrumentedApplication) bool {
+	return len(app.Spec.Applications) > 0
+}
+
+func (r *InstrumentedApplicationReconciler) detectLanguage(ctx context.Context, app *v1.InstrumentedApplication, labels map[string]string) error {
+	pod, err := r.choosePods(ctx, labels, app.Namespace)
+	if err != nil {
+		return err
+	}
+
+	langDetectionPod, err := r.createLangDetectionPod(pod, app)
+	if err != nil {
+		return err
+	}
+
+	err = r.Create(ctx, langDetectionPod)
+	return err
+}
+
+func (r *InstrumentedApplicationReconciler) choosePods(ctx context.Context, labels map[string]string, namespace string) (*corev1.Pod, error) {
```

Comment:
```suggestion
func (r *InstrumentedApplicationReconciler) choosePod(ctx context.Context, labels map[string]string, namespace string) (*corev1.Pod, error) {
```
Since it returns 1 pod and not a list